lua/rip-substitute/rg-operations.lua

```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
+---@param matches RipSubstituteMatch
+---@param range CmdRange
+local function getMatchesInRange(matches, range)
 	local rangeStartIdx, rangeEndIdx
-	if state.range then
-		for i = 1, #searchMatches do
-			local lnum = tonumber(searchMatches[i]:match("^(%d+):"))
-			local inRange = lnum >= state.range.start and lnum <= state.range.end_
-			if rangeStartIdx == nil and inRange then rangeStartIdx = i end
-			if rangeStartIdx and lnum > state.range.end_ then
-				rangeEndIdx = i - 1
-				break
-			end
+	for i, match in ipairs(matches) do
+		local inRange = match.row >= range.start and match.row <= range.end_
+		if rangeStartIdx == nil and inRange then rangeStartIdx = i end
+		if rangeStartIdx and match.row > range.end_ then
+			rangeEndIdx = i - 1
+			break
 		end
-		if rangeStartIdx == nil then return end -- no matches in range
-		searchMatches = vim.list_slice(searchMatches, rangeStartIdx, rangeEndIdx)
 	end
+	if rangeStartIdx == nil then return end -- no matches in range
+	return vim.list_slice(matches, rangeStartIdx, rangeEndIdx)
+end
 
-	state.matchCount = #searchMatches
-
+---@param matches RipSubstituteMatch
+---@param viewStartLnum number
+---@param viewEndLnum number
+---@return number|nil, number | nil
+local function getViewportRange(matches, viewStartLnum, viewEndLnum)
 	-- VIEWPORT: FILTER MATCHES
 	local viewStartIdx, viewEndIdx
-	for i = 1, #searchMatches do
-		local lnum = tonumber(searchMatches[i]:match("^(%d+):"))
-		if not viewStartIdx and lnum >= viewStartLnum and lnum <= viewEndLnum then
+	for i, match in ipairs(matches) do
+		if not viewStartIdx and match.row >= viewStartLnum and match.row <= viewEndLnum then
 			viewStartIdx = i
 		end
-		if viewStartIdx and lnum > viewEndLnum then
+		if viewStartIdx and match.row > viewEndLnum then
 			viewEndIdx = i - 1
 			break
 		end
 	end
-	if not viewStartIdx then return end -- no matches in viewport
-	if not viewEndIdx then viewEndIdx = #searchMatches end -- viewport is at end of file
+end
+
+---@param match RipSubstituteMatch
+---@param selected boolean
+---@param targetBuf number
+---@param incPreviewNs number
+function M.highlightReplacement(match, selected, targetBuf, incPreviewNs)
+	local hl_group = selected and config.incrementalPreview.hlGroups.currentMatch
+		or config.incrementalPreview.hlGroups.replacement
+	local ok, err = pcall(
+		function()
+			vim.api.nvim_buf_set_extmark(targetBuf, incPreviewNs, match.row, match.col, {
+				virt_text = {
+					{
+						match.replacementText,
+						hl_group,
+					},
+				},
+				virt_text_pos = "inline",
+				hl_mode = "replace",
+				strict = false,
+				conceal = match.matchedText,
+				end_col = match.col + #match.matchedText,
+				end_row = match.row,
+			})
+		end
+	)
+end
+
+--TODO: apparently some highlights that we want to keep get removed anyway-> todo comments
+--
+---@param match RipSubstituteMatch
+---@param selected boolean
+---@param targetBuf number
+---@param incPreviewNs number
```

Comment:
`targetBuf` and `incPreviewNs` do not need to be passed, since they can be accessed via `state`.